File: backend/app/schemas/analytics.py
"""Pydantic schemas for the analytics read model (Phase A1).

These describe the *read model* the charts dashboard consumes — deliberately
separate from the per-resource CRUD schemas. The dashboard must never receive
raw rows and compute trends in the browser, so these shapes are tailored to what
each chart draws:

- Food is pre-aggregated per day in SQL (a stacked area of macro totals).
- Weight and mood are returned as individual points (timestamp + value), *not*
  daily-averaged, so several same-day entries render as several dots on one x.
- Sleep is returned as intervals (start/end/duration) so a chart can show *when*
  each sleep happened, not just how long.

The day-detail response reuses the existing per-resource response schemas: the
drill-down shows the same entries the log pages do, with the same fields.
"""

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict

from app.schemas.food_entry import FoodEntryResponse
from app.schemas.mood_entry import MoodEntryResponse
from app.schemas.sleep_entry import SleepEntryResponse
from app.schemas.weight_entry import WeightEntryResponse


class FoodDayPoint(BaseModel):
    """One day's macro/calorie totals — aggregated in SQL, one point per day."""

    date: date
    calories: int
    protein_g: Decimal
    carb_g: Decimal
    fat_g: Decimal


class WeightPoint(BaseModel):
    """A single weigh-in. Kept per-entry (not daily-averaged) so multiple
    same-day weigh-ins plot as multiple dots sharing one x."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    measured_at: datetime
    weight: Decimal
    unit: str


class MoodPoint(BaseModel):
    """A single mood reading, kept per-entry for the same reason as weight."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    recorded_at: datetime
    mood_score: int


class SleepInterval(BaseModel):
    """One sleep as an interval, so a chart can show *when* it happened."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    started_at: datetime
    ended_at: datetime
    duration_minutes: int
    quality_score: int


class FoodSeries(BaseModel):
    items: list[FoodDayPoint]


class WeightSeries(BaseModel):
    items: list[WeightPoint]


class MoodSeries(BaseModel):
    items: list[MoodPoint]


class SleepSeries(BaseModel):
    items: list[SleepInterval]


class DayDetail(BaseModel):
    """Every entry for a single day, across all four trackers (Phase A3).

    "Which day" a timestamped entry belongs to is resolved in the user's
    timezone server-side; food already carries a calendar `entry_date`.
    """

    date: date
    food: list[FoodEntryResponse]
    weight: list[WeightEntryResponse]
    mood: list[MoodEntryResponse]
    sleep: list[SleepEntryResponse]

File: frontend/src/types/analytics.ts
// Mirror of the backend analytics read model (backend/app/schemas/analytics.py).
// These are the chart-ready shapes the dashboard consumes — the backend does
// the aggregation, the browser only draws.

import type { FoodEntry } from "~/types/food-entry";
import type { WeightEntry } from "~/types/weight-entry";
import type { MoodEntry } from "~/types/mood-entry";
import type { SleepEntry } from "~/types/sleep-entry";

// Numeric macros/weight arrive as strings (Decimal serialized as JSON string).

export type FoodDayPoint = {
  date: string; // ISO 8601 date, e.g. "2026-08-29"
  calories: number;
  protein_g: string;
  carb_g: string;
  fat_g: string;
};

export type WeightPoint = {
  id: number;
  measured_at: string; // ISO 8601 datetime (UTC)
  weight: string;
  unit: string;
};

export type MoodPoint = {
  id: number;
  recorded_at: string; // ISO 8601 datetime (UTC)
  mood_score: number;
};

export type SleepInterval = {
  id: number;
  started_at: string; // ISO 8601 datetime (UTC)
  ended_at: string; // ISO 8601 datetime (UTC)
  duration_minutes: number;
  quality_score: number;
};

export type FoodSeries = { items: FoodDayPoint[] };
export type WeightSeries = { items: WeightPoint[] };
export type MoodSeries = { items: MoodPoint[] };
export type SleepSeries = { items: SleepInterval[] };

// The day drill-down: every entry for one calendar day, across all trackers.
export type DayDetail = {
  date: string;
  food: FoodEntry[];
  weight: WeightEntry[];
  mood: MoodEntry[];
  sleep: SleepEntry[];
};
